File: 0057-insert-interval/0057-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        ArrayList<int[]> modified = new ArrayList();

        int i = 0;
        while (intervals.length > i
            && newInterval[0] > intervals[i][0]
            && newInterval[0] > intervals[i][1]) {
            modified.add(intervals[i]);

            i++;
        }

        if (intervals.length == i) {
            modified.add(newInterval);
            return modified.toArray(new int[modified.size()][2]);
        }

        if (newInterval[0] <= intervals[i][1]
            && newInterval[0] > intervals[i][0]) {
            newInterval[0] = intervals[i][0];
        }
        
        while (intervals.length > i && newInterval[1] >= intervals[i][0]) {
            i++;
        }

        if (0 < i && newInterval[1] < intervals[i - 1][1]) {
            newInterval[1] = intervals[i - 1][1];
        }

        modified.add(newInterval);

        while (intervals.length > i) {
            modified.add(intervals[i]);
            i++;
        }

        return modified.toArray(new int[modified.size()][2]);
    }
}

File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        i = 0
        n = len(intervals)
        
        # Add all intervals that come before the new interval
        while i < n and intervals[i][1] < newInterval[0]:
            res.append(intervals[i])
            i += 1
        
        # Merge overlapping intervals
        while i < n and intervals[i][0] <= newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[i][0])
            newInterval[1] = max(newInterval[1], intervals[i][1])
            i += 1
        
        # Add the merged interval
        res.append(newInterval)
        
        # Add remaining intervals
        while i < n:
            res.append(intervals[i])
            i += 1
        
        return res